[CORE-8532] Disallow certain schema evolution actions if a field appears in the partition spec

oleiman opened the pull request:
<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.


Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

General idea is as follows (quoted from the spec):

>Type promotion is not allowed for a field that is referenced by source-id or source-ids of a partition field if the partition transform would produce a different value after promoting the type.

In general, the type promotion rules preclude this, but date -> timestamp promotions are called out specifically as potentially rule violating, if the promoted field appears in the partition spec. So this diff prevents that by wiring the current partition spec into the compat module and adding a check to the validation path.

This PR also disallows removal of any data field which also appears in the current partition spec at the time of removal. This is technically outside the spec and needs followup ecosystem investigation, but it appears that dropping a column in this way can cause downstream catalog queries to fail. The safest thing at this stage is to simply disable the capability for now.

This PR also fixes a bug in `iceberg/schema_avro` whereby we would try to serialize `date_type` to an `AVRO_LONG` with logical type "date", which itself is restricted to `AVRO_INT`. Fix includes a unit test.

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v24.3.x
- [ ] v24.2.x
- [ ] v24.1.x

## Release Notes

* none

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

oleiman committed:
ice/compat: Plumb partition spec into compat checking

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

oleiman committed:
dt/dl/schema: Integration tests for partition source column type promotion

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

oleiman committed:
ice/schema: Correct logical type error in struct -> avro code

Logical DATE type should be bound to AVRO_INT only. Previously we used
AVRO_LONG, so the Avro lib would throw when exercising this code path.

Also adds a unit test for Iceberg types that are backed by Avro logicalTypes.

Note that timestamptz serialization is currently bugged, as it relies on an
extra non-standard field ('adjust-to-utc') in the Avro type description.
Technically this should appear in both micro precision timestamp types,
'false' for 'timestamp'; 'true' for 'timestamptz'. As written, both timestamp
types serialize to logical type timestamp-micros, with no way to distinguish
between the two in deser.

See https://iceberg.apache.org/spec/#avro for detail

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

oleiman committed:
ice/partition: Add lookup by source ID

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

oleiman committed:
dl/csm: Wire partition spec into compat checks

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

oleiman committed:
ice/compat: Make type_promoted an enum

To support special promotions that affect partition transforms

type_promoted:
  - no
  - yes
  - unless_partition

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

oleiman committed:
ice/compat: Disallow dropping partition fields from schema

Though this restriction does not appear in the spec, dropping a data column
that also appears in the table's partition spec can cause validation errors
in clients, downstream. It may be possible to avoid this by performing live
partition spec reconciliation inline with the schema update, but for now we
simply reject such an update out of hand.

Also slightly refactors to consolidate error checks in validate_schema_transform.

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--91ac725f-3299-4f00-a2a1-29e91cd44c70-->
<!--TEST RESULTS-->
<!--be459043-c482-4da4-80e7-4f4d3ee3aec8-->
<!--TEST RESULTS-->
<!--02419ed3-dc15-499a-baac-360a9591287d-->
<!--TEST RESULTS-->
<!--17b34ef1-0196-401e-a98f-fac502c0b324-->
<!--TEST RESULTS-->
<!--e6fd1bcb-a812-4f48-922b-d80e126385ff-->
<!--TEST RESULTS-->
<!--f84d382a-b3e8-4529-966d-6b5e49008965-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62039">build#62039</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.compaction_recovery_test.CompactionRecoveryTest.test_index_recovery | ducktape | https://buildkite.com/redpanda/redpanda/builds/62039#0195204f-d6c6-4734-b7e5-99f3c602c57d | FLAKY | 1/6 |
| rptest.tests.compaction_recovery_test.CompactionRecoveryTest.test_index_recovery | ducktape | https://buildkite.com/redpanda/redpanda/builds/62039#01952069-fe70-4baa-be48-2de4b36457af | FLAKY | 1/3 |
| rptest.tests.log_compaction_test.LogCompactionTest.compaction_stress_test.cleanup_policy=compact.delete.key_set_cardinality=1000.storage_compaction_key_map_memory_kb=10 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62039#01952069-fe68-40ed-8cbb-4d657447c169 | FLAKY | 1/2 |
| rptest.tests.log_compaction_test.LogCompactionTest.compaction_stress_test.cleanup_policy=compact.delete.key_set_cardinality=1000.storage_compaction_key_map_memory_kb=3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62039#01952069-fe6c-486a-ac2b-79b21753a320 | FLAKY | 1/3 |
| rptest.tests.log_compaction_test.LogCompactionTest.compaction_stress_test.cleanup_policy=compact.key_set_cardinality=1000.storage_compaction_key_map_memory_kb=3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62039#01952069-fe68-40ed-8cbb-4d657447c169 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62052">build#62052</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.log_compaction_test.LogCompactionTest.compaction_stress_test.cleanup_policy=compact.delete.key_set_cardinality=1000.storage_compaction_key_map_memory_kb=3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62052#019521ea-9063-4e47-a25f-59cb5322c60f | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62068">build#62068</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.archival_test.ArchivalTest.test_all_partitions_leadership_transfer.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/62068#019524ca-326e-4101-acf2-2837eb715b20 | FLAKY | 1/2 |
| rptest.tests.compaction_recovery_test.CompactionRecoveryTest.test_index_recovery | ducktape | https://buildkite.com/redpanda/redpanda/builds/62068#019524cd-f886-4dc0-9fb8-fd7dcb50f53b | FLAKY | 1/7 |
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape | https://buildkite.com/redpanda/redpanda/builds/62068#019524ca-3270-41ea-b4e3-fcb846f59b0f | FLAKY | 1/2 |
| rptest.tests.e2e_shadow_indexing_test.ShadowIndexingWhileBusyTest.test_create_or_delete_topics_while_busy.short_retention=True.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/62068#019524cd-f885-48ec-99bc-cacb1c6b73b1 | FLAKY | 1/2 |
| rptest.tests.full_disk_test.FullDiskReclaimTest.test_full_disk_triggers_gc | ducktape | https://buildkite.com/redpanda/redpanda/builds/62068#019524ca-3270-41ea-b4e3-fcb846f59b0f | FLAKY | 1/2 |
| rptest.tests.log_compaction_test.LogCompactionTest.compaction_stress_test.cleanup_policy=compact.key_set_cardinality=1000.storage_compaction_key_map_memory_kb=3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62068#019524cd-f884-4f87-85c1-e01181e315da | FLAKY | 1/2 |
| rptest.tests.partition_movement_test.SIPartitionMovementTest.test_shadow_indexing.num_to_upgrade=0.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/62068#019524cd-f884-4f87-85c1-e01181e315da | FLAKY | 1/2 |
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic | ducktape | https://buildkite.com/redpanda/redpanda/builds/62068#019524cd-f884-4f87-85c1-e01181e315da | FLAKY | 1/3 |
| rptest.tests.write_caching_fi_e2e_test.WriteCachingFailureInjectionE2ETest.test_crash_all_with_consumer_group | ducktape | https://buildkite.com/redpanda/redpanda/builds/62068#019524ca-3270-41ea-b4e3-fcb846f59b0f | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62129">build#62129</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.datalake.datalake_dlq_test.DatalakeDLQTest.test_dlq_table_for_mixed_records.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.TRINO.catalog_type=CatalogType.REST_JDBC | ducktape | https://buildkite.com/redpanda/redpanda/builds/62129#01952ce8-52f5-4446-80fb-7bf316f100c0 | FLAKY | 1/2 |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_partition_spec_evo.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.SPARK.use_partition_spec=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/62129#01952ce8-52f5-4446-80fb-7bf316f100c0 | FAIL | 0/20 |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_partition_spec_evo.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.SPARK.use_partition_spec=True | ducktape | https://buildkite.com/redpanda/redpanda/builds/62129#01952ce8-52f2-4457-93c8-391ae45a5c02 | FAIL | 0/20 |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_partition_spec_evo.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.TRINO.use_partition_spec=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/62129#01952ce8-52f3-4dc6-9bcc-e2c41c46417d | FAIL | 0/20 |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_partition_spec_evo.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.TRINO.use_partition_spec=True | ducktape | https://buildkite.com/redpanda/redpanda/builds/62129#01952ce8-52f4-4c6c-a825-c9dec126b27e | FAIL | 0/20 |
| rptest.tests.internal_topic_protection_test.InternalTopicProtectionLargeClusterTest.test_consumer_offset_topic | ducktape | https://buildkite.com/redpanda/redpanda/builds/62129#01952ce8-52f5-4446-80fb-7bf316f100c0 | FLAKY | 1/2 |
| rptest.transactions.producers_api_test.ProducersAdminAPITest.test_producers_state_api_during_load | ducktape | https://buildkite.com/redpanda/redpanda/builds/62129#01952ce8-52f5-4446-80fb-7bf316f100c0 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62149">build#62149</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.e2e_shadow_indexing_test.ShadowIndexingWhileBusyTest.test_create_or_delete_topics_while_busy.short_retention=True.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/62149#0195311a-531b-4e89-8e2f-808798ce7916 | FLAKY | 1/2 |
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic | ducktape | https://buildkite.com/redpanda/redpanda/builds/62149#0195311a-531a-4c74-8fb1-c503c0b4f272 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62177">build#62177</a></summary>


| test_id                                                                                                                                                                                                                                           | test_kind   | job_url                                                                                   | test_status   | passed   |
|:--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.e2e_shadow_indexing_test.ShadowIndexingWhileBusyTest.test_create_or_delete_topics_while_busy.short_retention=True.cloud_storage_type=CloudStorageType.S3                                                                             | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62177#0195368d-de5f-4685-a8c9-e2693ade4b48 | FLAKY         | 1/2      |
| rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=False.mixed_versions=True.with_tiered_storage=True.with_iceberg=False.with_chunked_compaction=False.cloud_storage_type=CloudStorageType.S3 | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62177#0195368d-de59-4e56-8dba-d6fc18b90541 | FLAKY         | 1/2      |
</details>
<!--/body-->

oleiman committed:
dl/dt: Test for dropping partition field from table

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

oleiman commented:
/ci-repeat 1
skip-redpanda-build
skip-units

oleiman commented:
CI Failures:
- [ducktape-debug](https://buildkite.com/redpanda/redpanda/builds/62089#019526eb-3961-4fe8-860a-d2b448423a73), [ducktape-release](https://buildkite.com/redpanda/redpanda/builds/62089#01952706-1f68-4244-bcff-042188c90147) - both look like infra

andrwng commented on src/v/iceberg/schema_avro.cc:
```diff
@@ -79,7 +79,7 @@ struct avro_primitive_type_visitor {
     }
     avro::Schema operator()(const date_type&) {
-        auto ret = avro::LongSchema();
+        auto ret = avro::IntSchema();
```
Oof nice catch

andrwng commented on src/v/iceberg/compatibility_types.h:
```diff
@@ -23,7 +23,12 @@ enum class compat_errc {
+enum class type_promoted {
+    no,
+    yes,
+    unless_partition,
```
nit: can you add a comment explaining what this is

oleiman replied:
done

andrwng commented on src/v/iceberg/compatibility.h:
```diff
@@ -31,6 +32,9 @@ namespace iceberg {
    @return The result of the type check:
              - type_promoted::yes - if src -> dest is a valid type promotion
                e.g. int -> long
+             - type_promoted::unless_partition - if src -> dest is a valid type
```
Trying to wrap my head around the naming here -- can you use "unless partition" in a complete sentence that describes it? 😅

andrwng replied:
Ah I guess in the context of its usage, the "unless" part comes in, "yes the type promotion is valid unless it changes the partition key"

As an output of `check_types()` I think it takes less context to understand if the name is just some `changes_partition`, and then it'd feel pretty intuitive that the caller checks if the field is a part of the partition key. WDYT?

oleiman replied:
yeah it is horrible. was hoping for another suggestion, so `changes_partition` wfm!

oleiman replied:
renamed

andrwng commented on src/v/iceberg/tests/compatibility_test.cc:
```diff
@@ -540,6 +547,7 @@ static const std::vector<struct_evolution_test_case> valid_cases{
                  && dst_nested.fields.empty()
                  && removed(*src_nested.fields.back());
       },
+    .pspec = "(nested)", // TODO(oren): seems wrong frankly, should this fail?
```
>The source columns, selected by ids, must be a primitive type and cannot be contained in a map or list, but may be nested in a struct.

Hmm yea this is interesting. Maybe not in this PR, but feels like we should treat it like we would a missing field when resolving https://github.com/redpanda-data/redpanda/blob/5fe9fc9c8c7f36cd59c427190a0aeea13c6a61dd/src/v/iceberg/partition.cc#L38-L40

oleiman replied:
ohh, i see. so a struct or map or whatever isn't even a valid source column? I don't think partition spec resolution is detecting that at all then. See L448 in this file e.g.... seems like a bug.

oleiman replied:
split if off in case there's more to discuss. https://github.com/redpanda-data/redpanda/pull/25144 https://github.com/redpanda-data/redpanda/commit/62c61f7f410087e565ed2c513a310255bceea5d0

andrwng commented on src/v/iceberg/tests/compatibility_test.cc:
```diff
@@ -998,10 +1038,16 @@ TEST_P(AnnotateStructTest, AnnotationWorksAndDetectsStructuralErrors) {
-            ? fmt::format("Unexpected error: {}", annotate_res.error())
-            : fmt::format("Expected {}", err().error()));
+    if (
+      err().has_error()
+      && err().error() == schema_evolution_errc::partition_spec_conflict) {
```
Hmm why wouldn't the test just set no err() value?

oleiman replied:
rough edge from using the same suite of test cases for everything. i will revisit to smooth out.

oleiman replied:
refactored test cases a bit

andrwng commented on tests/rptest/tests/datalake/schema_evolution_test.py:
```diff
@@ -769,3 +780,82 @@ def test_old_schema_writer(self, cloud_storage_type, query_engine,
+        self.logger.debug(
+            f"Schema evolution {'fails' if use_partition_spec else 'succeeds'} if the date field is {'not ' if not use_partition_spec else ''}referenced in the partition spec"
+        )
+        pspec = tc.partition_spec if use_partition_spec else None
```
nit: would it make sense to move this condition to L830?

oleiman replied:
done

andrwng commented on tests/rptest/tests/datalake/schema_evolution_test.py:
```diff
@@ -769,3 +780,82 @@ def test_old_schema_writer(self, cloud_storage_type, query_engine,
+            partition_spec="(ts)",
+        )
+
+        self.logger.debug(
```
nit: maybe it's just me, feels like this would be easier to parse/verify as a reader if just

```
if use_partition_spec:
  log...
else:
  log...
```

oleiman replied:
oh yeah, no objections here. when I strap on my python hat I become a 🏌️‍♂️

oleiman replied:
done

andrwng commented on tests/rptest/tests/datalake/schema_evolution_test.py:
```diff
@@ -769,3 +780,82 @@ def test_old_schema_writer(self, cloud_storage_type, query_engine,
+                                   mode=ProducerType.AVRO)
+
+            if use_partition_spec:
+                tc.initial_schema.check_table_schema(dl, self.table_name,
+                                                     query_engine)
```
nit: maybe worth checking for records in the DLQ table?

oleiman replied:
oh yeah, great idea. when i wrote these initially i think the dlq was still WIP. would be great to add that to all the tests here.

oleiman replied:
done

andrwng commented on tests/rptest/tests/datalake/schema_evolution_test.py:
```diff
@@ -769,3 +780,82 @@ def test_old_schema_writer(self, cloud_storage_type, query_engine,
+        query_engine=QUERY_ENGINES,
+        use_partition_spec=[True, False],
+    )
+    def test_partition_spec_evo(self, cloud_storage_type, query_engine,
```
Not related to this PR, but it's worth trying these out with a a real catalog instead of just filesystem_catalog

oleiman replied:
agree

andrwng commented on src/v/iceberg/compatibility.cc:
```diff
@@ -10,6 +10,7 @@
 
 #include "iceberg/compatibility.h"
 
+#include "iceberg/compatibility_types.h"
```
>ice/compat: Make field removal subject to partition spec

Can you add more context here?

oleiman replied:
oof, good catch

oleiman replied:
done

andrwng commented on src/v/iceberg/compatibility.cc:
```diff
@@ -141,8 +142,12 @@ schema_transform_state add_field(const nested_field& f) {
+remove_field(const nested_field& f, const partition_spec& pspec) {
     f.set_evolution_metadata(nested_field::removed::yes);
+    if (pspec.get_field(f.id) != nullptr) {
+        return {.n_removed_partition_fields = 1};
```
Should this also set n_removed = 1? Maybe worth a comment explaining why

oleiman replied:
done. comment in header

andrwng commented on src/v/iceberg/compatibility_types.h:
```diff
@@ -55,6 +55,7 @@ struct schema_transform_state {
     size_t n_added{0};
     size_t n_promoted{0};
+    size_t n_removed_partition_fields{0};
     size_t total() { return n_removed + n_added + n_promoted; }
```
Should this include n_removed_partition_fields? Maybe worth a comment explaining why

oleiman replied:
yeah i waffled a bit. if we remove partition fields that's an error right now, but in the future it might feed back up to some recovery action. good call.

oleiman replied:
for posterity:
- increment `n_removed` in the removal routine
- don't include `n_removed_partition` in sum (double count).
- comment on `state::total`

andrwng commented on tests/rptest/tests/datalake/schema_evolution_test.py:
```diff
@@ -439,6 +439,51 @@ class EvolutionTestCase(NamedTuple):
+            generate_record=lambda x: {
+                "verifier_string": f"verify-{x}",
+            },
+            spark_table=[
+                ('verifier_string', 'string'),
+            ],
+            trino_table=[
+                ('verifier_string', 'varchar'),
+            ],
```
Hmm can you remind me how these are used? We _don't_ expect the drop to land in the table, right?

oleiman replied:
i would think of the individual `GenericSchema` objects as describing a schema, what it should look like as a table description, and providing an example record. Then the test case describes a change. And the test function itself asserts, in this case, that the table didn't change (i.e. we assert on the original schema).

But yeah, perhaps it's more clear to leave these blank when they won't be used.

oleiman replied:
misunderstood your question. no, these are not needed - removed.

andrwng reviewed (commented):
Looking good! Mostly a bunch of nits and questions

oleiman committed:
ice/tests: Refactor compatibility_test test cases

Specify either an annotate error or a validate error. This simplifies some of
the conditional logic around instantiating test suites. No functional changes.

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/62129-->
<!--eb12bcf6-4133-46be-97f3-17f49a9f49de-->
<!--9dbe9305-9a93-4f37-adb3-13eba11288be-->
<!--d03c3e24-d2ae-4acb-a0e4-7eeb7c659642-->
<!--f7e5ca2a-7926-465a-9ee9-19898bc93da5-->
<!--title-->
## Retry command for [Build#62129](https://buildkite.com/redpanda/redpanda/builds/62129)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```


/ci-repeat 1
tests/rptest/tests/datalake/schema_evolution_test.py::SchemaEvolutionE2ETests.test_partition_spec_evo@{"cloud_storage_type":1,"query_engine":"spark","use_partition_spec":false}
tests/rptest/tests/datalake/schema_evolution_test.py::SchemaEvolutionE2ETests.test_partition_spec_evo@{"cloud_storage_type":1,"query_engine":"spark","use_partition_spec":true}
tests/rptest/tests/datalake/schema_evolution_test.py::SchemaEvolutionE2ETests.test_partition_spec_evo@{"cloud_storage_type":1,"query_engine":"trino","use_partition_spec":true}
tests/rptest/tests/datalake/schema_evolution_test.py::SchemaEvolutionE2ETests.test_partition_spec_evo@{"cloud_storage_type":1,"query_engine":"trino","use_partition_spec":false}
```
<!--/body-->

rockwotj reviewed (dismissed):
Very nice

oleiman commented:
[force push](https://github.com/redpanda-data/redpanda/compare/fbb780d75fdb3944bbbbd0c6d6b05faa517e1b7e..035c4e38ebba2eb582fc664bee8dc0f35b364e25) typos

oleiman committed:
dt/dl/schema: Use DLQ rather than timeout for invalid schema checks

oleiman commented:
/ci-repeat 1

oleiman commented:
/ci-repeat 1
release
skip-redpanda-build
skip-units
